### ISSUE TITLE
Add Gather methods for Vector<T>

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/runtime#14522]
Vector Gathers are a commonly used paradigm in vector computation - where a gather involves constructing a vector from non-contiguous elements of a source array.  Gather is supported on AVX2, but even on targets that do not directly support it, such a method would be useful, in order to avoid the overhead of having to use an array to hold the elements prior to contructing the Vector.

VectorGather should perhaps be declared on the static Vector class, something like:
    `public static Vector<T> VectorGather(T[] source, Vector<U> indexVector) where T,U : struct`

Initially, I think it would be reasonable to support only same/size convert and gather – but eventually we may want to consider mixed size (e.g. a gather to a Vector<Double> using the lower values of a Vector<Int32>.  AVX2 has a pretty rich mixture of these.

@mellinoe what do you think?

**Comment by tannergooding** [comment]
Going to close this for now. This API needs some more thought about how gathering would work with when exposed as a general-purpose API.

If someone feels strongly about this, please feel free to open a new issue. Our API review process is detailed [here](https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md) with a "good example" being listed under step 1. Ideally the new issue would provide many of the sections given, but would at a minimum provide the base rationale and proposed API surface sections.

For this API in particular, the exposure of the System.Runtime.Intrinsic types and the methods for converting to/from `Vector<T>` should unblock users who immediately need this functionality.
